web/packages/teleterm/src/ui/services/deepLinks/deepLinksService.ts

```diff
@@ -88,13 +88,70 @@ export class DeepLinksService {
         break;
       }
       case '/authenticate_web_device': {
-        // TODO (avatus): add authenticate device web confirmation dialog
-        // this case is a stub and will not be reached
+        const id = result.url?.searchParams?.id;
+        const token = result.url?.searchParams?.token;
+        await this.askAuthorizeDeviceTrust(result.url, id, token);
         break;
       }
     }
   }
 
+  private async askAuthorizeDeviceTrust(
+    url: DeepURL,
+    id: string,
+    token: string
+  ): Promise<void> {
+    const rootClusterId = url.hostname;
+    const clusterAddress = url.host;
+    const prefill = {
+      clusterAddress,
+      username: url.username,
+    };
+
+    const rootClusterUri = routing.getClusterUri({ rootClusterId });
+    const rootCluster = this.clustersService.findCluster(rootClusterUri);
+
+    if (!rootCluster) {
+      const { canceled } = await new Promise<{ canceled: boolean }>(resolve => {
+        this.modalsService.openRegularDialog({
+          kind: 'cluster-connect',
+          clusterUri: undefined,
+          reason: undefined,
+          prefill,
+          onSuccess: () => resolve({ canceled: false }),
+          onCancel: () => resolve({ canceled: true }),
+        });
+      });
+
+      if (canceled) {
+        return;
+      }
+    }
+
+    const { isAtDesiredWorkspace } =
+      await this.workspacesService.setActiveWorkspace(rootClusterUri, prefill);
+
+    if (!isAtDesiredWorkspace) {
+      return;
+    }
+
+    this.modalsService.openImportantDialog({
+      kind: 'device-trust-authorize',
+      onCancel: () => console.log('cancelled'),
+      onConfirm: async () => {
+        await this.clustersService.authenticateWebDevice(rootClusterUri, {
+          id,
+          token,
+        });
+        // open url to confirm the token. This endpoint verifies the token and "upgrades"
+        // the web session and redirects to "/web"
+        window.open(
+          `https://${url.host}/webapi/devices/webconfirm?id=${id}&token=${token}`
```

Comment:
I think we're better off using `rootCluster.proxyHost` instead of `url.host`, just because `url.host` comes from the user and `rootCluster.proxyHost` is guaranteed to come from a trusted source.

Reply:
That's fair. The openExternal is set to only open safe URLs so it's not the worst thing to try and open something here as only GitHub and the root are able to be opened anyway but, the suggestion will make it super safe so that's ok with me

Reply:
Ah, yeah, that's true. I guess it just comes down to the principle of always preferring validated/parsed/"correct" data over one that came from an untrusted source.